datafusion/optimizer/src/propagate_empty_relation.rs

```diff
@@ -0,0 +1,331 @@
+            | LogicalPlan::Filter(_)
+            | LogicalPlan::Window(_)
+            | LogicalPlan::Sort(_)
+            | LogicalPlan::SubqueryAlias(_)
+            | LogicalPlan::Repartition(_)
+            | LogicalPlan::Limit(_) => match empty_child(&optimized_children_plan)? {
+                Some(empty) => Ok(empty),
+                None => Ok(optimized_children_plan),
+            },
+            LogicalPlan::CrossJoin(_) => {
+                let (left_empty, right_empty) = binary_plan_children_is_empty(plan)?;
+                if left_empty || right_empty {
+                    Ok(LogicalPlan::EmptyRelation(EmptyRelation {
+                        produce_one_row: false,
+                        schema: plan.schema().clone(),
+                    }))
+                } else {
+                    Ok(optimized_children_plan)
+                }
+            }
+            LogicalPlan::Join(join) => {
+                // TODO: For Join, more join type need to be careful:
+                // For LeftOuter/LeftSemi/LeftAnti Join, only the left side is empty, the Join result is empty.
+                // For LeftSemi Join, if the right side is empty, the Join result is empty.
+                // For LeftAnti Join, if the right side is empty, the Join result is left side(should exclude null ??).
+                // For RightOuter/RightSemi/RightAnti Join, only the right side is empty, the Join result is empty.
+                // For RightSemi Join, if the left side is empty, the Join result is empty.
+                // For RightAnti Join, if the left side is empty, the Join result is right side(should exclude null ??).
+                // For Full Join, only both sides are empty, the Join result is empty.
+                // For LeftOut/Full Join, if the right side is empty, the Join can be eliminated with a Projection with left side
+                // columns + right side columns replaced with null values.
+                // For RightOut/Full Join, if the left side is empty, the Join can be eliminated with a Projection with right side
+                // columns + left side columns replaced with null values.
+                if join.join_type == JoinType::Inner {
+                    let (left_empty, right_empty) = binary_plan_children_is_empty(plan)?;
+                    if left_empty || right_empty {
+                        Ok(LogicalPlan::EmptyRelation(EmptyRelation {
+                            produce_one_row: false,
+                            schema: plan.schema().clone(),
+                        }))
+                    } else {
+                        Ok(optimized_children_plan)
+                    }
+                } else {
+                    Ok(optimized_children_plan)
+                }
+            }
+            LogicalPlan::Union(union) => {
+                let (left_empty, right_empty) = binary_plan_children_is_empty(plan)?;
+                if left_empty && right_empty {
```

Comment:
I'm not sure whether DataFusion's Union is binary or can have more than 2 inputs.